umpost/um2netcdf.py

```diff
@@ -99,46 +126,193 @@ def convert_proleptic(time):
-                coord.guess_bounds()
-        else:
-            # For length 1, assume it's global. guess_bounds doesn't work in this case
-            if coord.name() == 'latitude':
-                if not coord.has_bounds():
-                    coord.bounds = np.array([[-90., 90.]])
-            elif coord.name() == 'longitude':
-                if not coord.has_bounds():
-                    coord.bounds = np.array([[0., 360.]])
+def fix_lat_coord_name(lat_coordinate, grid_type, dlat):
+    """
+    Add a 'var_name' attribute to a latitude coordinate object
+    based on the grid it lies on.
 
-    lat = cube.coord('latitude')
+    NB - Grid spacing dlon only refers to variables on the main
+    horizontal grids, and not the river grid.
 
-    # Force to double for consistency with CMOR
-    lat.points = lat.points.astype(np.float64)
-    _add_coord_bounds(lat)
-    lon = cube.coord('longitude')
-    lon.points = lon.points.astype(np.float64)
-    _add_coord_bounds(lon)
-
-    lat = cube.coord('latitude')
-    if len(lat.points) == 180:
-        lat.var_name = 'lat_river'
-    elif (lat.points[0] == -90 and grid_type == 'EG') or \
-         (np.allclose(-90.+0.5*dlat, lat.points[0]) and grid_type == 'ND'):
-        lat.var_name = 'lat_v'
+    Parameters
+    ----------
+    lat_coordinate: coordinate object from iris cube (edits in place).
+    grid_type: (string) model horizontal grid type.
+    dlat: (float) meridional spacing between latitude grid points.
+    """
+
+    if lat_coordinate.name() != LATITUDE:
+        raise ValueError(
+                f"Wrong coordinate {lat_coordinate.name()} supplied. "
+                f"Expected {LATITUDE}."
+            )
+
+    if is_lat_river_grid(lat_coordinate.points):
+        lat_coordinate.var_name = VAR_NAME_LAT_RIVER
+    elif is_lat_v_grid(lat_coordinate.points, grid_type, dlat):
+        lat_coordinate.var_name = VAR_NAME_LAT_V
+    else:
+        lat_coordinate.var_name = VAR_NAME_LAT_STANDARD
+
+
+def fix_lon_coord_name(lon_coordinate, grid_type, dlon):
+    """
+    Add a 'var_name' attribute to a longitude coordinate object
+    based on the grid it lies on.
+
+    NB - Grid spacing dlon only refers to variables on the main
+    horizontal grids, and not the river grid.
+
+    Parameters
+    ----------
+    lon_coordinate: coordinate object from iris cube (edits in place).
+    grid_type: (string) model horizontal grid type.
+    dlon: (float) zonal spacing between longitude grid points.
+    """
+
+    if lon_coordinate.name() != LONGITUDE:
+        raise ValueError(
+                f"Wrong coordinate {lon_coordinate.name()} supplied. "
+                f"Expected {LATITUDE}."
+            )
+
+    if is_lon_river_grid(lon_coordinate.points):
+        lon_coordinate.var_name = VAR_NAME_LON_RIVER
+    elif is_lon_u_grid(lon_coordinate.points, grid_type, dlon):
+        lon_coordinate.var_name = VAR_NAME_LON_U
+    else:
+        lon_coordinate.var_name = VAR_NAME_LON_STANDARD
+
```

Comment:
Given these two functions have an almost identical structure, is it worth considering refactoring them as follows:
- A private helper function that does the actual work;
- Turn the existing functions into wrappers that call the private functions with the expecyed coordinate name, and comparison options for that coordinate?

Reply:
I've put together a trial of this - let me know if it's along the lines of what you were thinking! The first attempt was using a private function:
```python

def _fix_horizontal_coord_name(coordinate, grid_type, grid_spacing, 
                               river_grid_check,
                               river_grid_name, staggered_grid_check,
                               staggered_name, base_name):
    if river_grid_check(coordinate.points):
        coordinate.var_name = river_grid_name
    elif staggered_grid_check(coordinate, grid_type, grid_spacing):
        coordinate.var_name = staggered_name
    else:
        coordinate.var_name = base_name



def fix_lat_coord_name(lat_coordinate, grid_type, dlat):
    _fix_horizontal_coord_name(coordinate = lat_coordinate,
                               grid_type=grid_type,
                               grid_spacing=dlat,
                               river_grid_check=is_lat_river,
                               river_grid_name=VAR_NAME_LAT_RIVER,
                               staggered_grid_check=is_lat_v_grid,
                               staggered_name=VAR_NAME_LAT_V,
                               base_name=VAR_NAME_LAT_STANDARD)
```

And another option was to put together a dictionary holding the specified checks:
```python
HORIZONTAL_GRID_NAMING_DATA = {
    LATITUDE: {
        "river_grid_check": is_lat_river_grid,
        "river_grid_name": VAR_NAME_LAT_RIVER,
        "staggered_name": VAR_NAME_LAT_V,
        "base_name":  VAR_NAME_LAT_STANDARD,
        "staggered_grid_check": is_lat_v_grid
    },
    LONGITUDE: {
        "river_grid_check": is_lon_river_grid,
        "river_grid_name": VAR_NAME_LON_RIVER,
        "staggered_grid_check": is_lon_u_grid,
        "staggered_name": VAR_NAME_LON_U,
        "base_name":  VAR_NAME_LAT_STANDARD,
    }
}


def fix_latlon_coord_name(coordinate, grid_type, grid_spacing):
    coord_name = coordinate.name

    if HORIZONTAL_GRID_NAMING_DATA["river_grid_check"](coordinate.points):
        coordinate.var_name = HORIZONTAL_GRID_NAMING_DATA["river_grid_name"]
    elif HORIZONTAL_GRID_NAMING_DATA["staggered_grid_check"](coordinate.points, grid_type, grid_spacing):
        coordinate.var_name = HORIZONTAL_GRID_NAMING_DATA["staggered_name"]
    else:
        coordinate.var_name = HORIZONTAL_GRID_NAMING_DATA["base_name"]
```
In terms of readability, I think I find the separate tests a bit easier to follow, though it is some amount of duplicate code. Happy to hear other opinions and ideas though!

Reply:
Are these experiments on other branches?

While the code & tests work here (with a bit of duplication), refining the code can be deferred in favour of modularisation & test coverage. @blimlim What are your thoughts on adding an neatening task to issue #27, linking it to this discussion & returning to it later?

Reply:
That sounds like a good idea, I've added a task to #27.  The two experiments are up on the following branches:
https://github.com/ACCESS-NRI/um2nc-standalone/tree/55/refactor-fix_latlon_coord-private-name-func
https://github.com/ACCESS-NRI/um2nc-standalone/tree/55/refactor-fix_latlon_coord-dictionary-name-func

Reply:
Deferred action sounds like a good idea.

My original thought was like your option 2, but even more general - each coordinate could have an iterable of (check, name) tuples that you go through until you hit a match. The last tuple in the iterable would then be `(True, <default name>)`.